src/encoding.rs

```diff
@@ -762,20 +815,38 @@ pub mod bytes {
     where
         B: Buf,
     {
+        // Append `s` to `v` using a `memcpy`.
+        // Safety: `v` must have enough capacity to accommodate `s`.
+        unsafe fn copy_bytes_to_vec(v: &mut Vec<u8>, s: &[u8]) {
+            let v_len = v.len();
+            let s_len = s.len();
+            let dst = v.as_mut_slice().get_unchecked_mut(v_len) as *mut u8;
+            ::std::ptr::copy_nonoverlapping::<u8>(&s[0] as *const u8, dst, s_len);
+            v.set_len(v_len + s_len);
+        }
+
         check_wire_type(WireType::LengthDelimited, wire_type)?;
         let len = decode_varint(buf)?;
         if len > buf.remaining() as u64 {
             return Err(DecodeError::new("buffer underflow"));
         }
 
         let mut remaining = len as usize;
+        value.reserve(remaining);
+
         while remaining > 0 {
             let len = {
                 let bytes = buf.bytes();
                 debug_assert!(!bytes.is_empty(), "Buf::bytes returned empty slice");
                 let len = min(remaining, bytes.len());
                 let bytes = &bytes[..len];
-                value.extend_from_slice(bytes);
+                unsafe {
+                    debug_assert!(
+                        value.capacity() >= bytes.len(),
+                        "Should be ensured by the `reserve` statement above",
+                    );
+                    copy_bytes_to_vec(value, bytes);
```

Comment:
Were you able to isolate this change and get improved numbers?  As I understand it the savings here is one removed bounds check per loop iteration on the `value` buffer.  Otherwise, `extend_from_slice` should boil down to the same call to `copy_nonoverlapping` internally.

Importantly, when the buf type is `Vec` or `Bytes` (e.g., almost always), this loop will only be iterated once, since `buf.bytes()` will return the full buffer.  As such I'd be surprised if this meaningfully moved the needle, and if it did I'd suspect it came down to inlining changes more than bounds checks.

Reply:
> Were you able to isolate this change and get improved numbers?

I didn't get accurate numbers for each change, but I did see an improvement for each one (and tried lots of things which gave no improvement and dumped them). I'll try and get something more precise.

> Otherwise, extend_from_slice should boil down to the same call to copy_nonoverlapping internally.

Because `extend_from_slice` is generic and there is a lot of indirection in the implementation I'm not sure whether we hit the fast path or not.

> Importantly, when the buf type is Vec or Bytes (e.g., almost always)

This is not the case for us - we get a buffer from grpc and it can come a piece at a time.